### PR TITLE
[PATCH v2] Added DPDK driver HW timestamp Mbuf extention support.

### PR DESCRIPTION
**Pull request by rivasiv**
#131 

- Added DPDK HW 64-bit timestamp;
- Moved ts_all ts_ptp handling inside cycle to allow each pkt be marked with it's own timestamp.

### CODE REVIEW

**Comment by rivasiv** [comment]
@MatiasElo Could you take a look when would have a minute ?

File: include/odp/api/spec/packet_io.h [hunk 1]
```diff
@@ -393,6 +393,10 @@ typedef union odp_pktin_config_opt_t {
 		/** Drop packets with a SCTP error on packet input */
 		uint64_t drop_sctp_err : 1;
 
+		/**Apply hardware timestamp on pkt input. Should be supported 
+		 * by NIC driver. */ 
+		uint64_t ts_hw : 1;
+
```
Comment (MatiasElo) [inline include/odp/api/spec/packet_io.h:399]: All ODP API changes are done through the linux-generic project (https://github.com/OpenDataPlane/odp). At a glance this addition seems unnecessary. The existing `ts_all` field could be used instead and the implementation can decide to use SW/HW based on the HW capabilities.

File: platform/linux-dpdk/odp_packet_dpdk.c [hunk 2]
```diff
@@ -136,6 +136,14 @@ const pktio_if_ops_t * const _odp_pktio_if_ops[]  = {
 	NULL
 };
 
+/* Timestamp offset in rte mbuf.*/
+static int rte_mbuf_hw_ts_dynfield_offset = -1;
```
Comment (MatiasElo) [inline platform/linux-dpdk/odp_packet_dpdk.c:140]: This value should be stored inside `pkt_dpdk_t` to enable process mode operation.

File: platform/linux-dpdk/odp_packet_dpdk.c [hunk 3]
```diff
@@ -285,6 +293,17 @@ static int dpdk_setup_eth_dev(pktio_entry_t *pktio_entry)
...
+	if (pktio_entry->s.config.pktin.bit.ts_hw) {
+		rx_offloads |= DEV_RX_OFFLOAD_TIMESTAMP;
+
+		rte_mbuf_dyn_rx_timestamp_register(&rte_mbuf_hw_ts_dynfield_offset, NULL);
+		if (rte_mbuf_hw_ts_dynfield_offset < 0) {
+			//Disable HW mbuf timestamping since we would not be able to read it.
+			pktio_entry->s.config.pktin.bit.ts_hw = 0;
+			ODP_ERR("rte_mbuf_dyn_rx_timestamp_register failed. ts_hw disabled!\n");
+		}
+	}
```
Comment (MatiasElo) [inline platform/linux-dpdk/odp_packet_dpdk.c:305]: As I mentioned in the API change comment, perhaps a better solution would be to check if HW timestamping is supported, and if not, fall back to using SW timestamps.

File: platform/linux-dpdk/odp_packet_dpdk.c [hunk 4]
```diff
@@ -285,6 +293,17 @@ static int dpdk_setup_eth_dev(pktio_entry_t *pktio_entry)
 	if (pktio_entry->s.config.pktin.bit.tcp_chksum)
 		rx_offloads |= DEV_RX_OFFLOAD_TCP_CKSUM;
 
+	if (pktio_entry->s.config.pktin.bit.ts_hw) {
+		rx_offloads |= DEV_RX_OFFLOAD_TIMESTAMP;
+
+		rte_mbuf_dyn_rx_timestamp_register(&rte_mbuf_hw_ts_dynfield_offset, NULL);
```
Comment (MatiasElo) [inline platform/linux-dpdk/odp_packet_dpdk.c:299]: ODP-DPDK supports both current DPDK LTS releases (v19.11, v20.11) and this API is not yet available in DPDK v19.11.

Comment (MatiasElo) [reply platform/linux-dpdk/odp_packet_dpdk.c:299]: What is also problematic with this API is that it doesn't specify anything about the resolution or units of timestamps, which would be required to implement `odp_pktio_ts_res()`.

File: include/odp/api/spec/packet_io.h (hunk 1 above)
Comment (rivasiv) [reply include/odp/api/spec/packet_io.h:399]: That of corse may be hidden under ts_all. In that case user is not able to select which one to use if there are two options. And some priorities on selection needs to be added. Which would not handle all the cases right - only behaving by some algorithm. We loosing flexibility. I think it's possible to add the same API to linux-generic withSOF_TIMESTAMPING_RX_HARDWARE support.

File: platform/linux-dpdk/odp_packet_dpdk.c (hunk 3 above)
Comment (rivasiv) [reply platform/linux-dpdk/odp_packet_dpdk.c:305]: In general that would work for "my" case. But in this variant we loosing flexibility. And on processing stage not really sure from where TS is going from .

File: platform/linux-dpdk/odp_packet_dpdk.c [hunk 5]
```diff
@@ -930,6 +946,19 @@ int _odp_input_pkts(pktio_entry_t *pktio_entry, odp_packet_t pkt_table[], int nu
 				continue;
 			}
 		}
+
+		/*Moved this into pkt processing cycle since we want to have unique TS for each ptk, rather than one for group*/
+		if (pktio_entry->s.config.pktin.bit.ts_all ||
+			pktio_entry->s.config.pktin.bit.ts_ptp) {
+			ts_val = odp_time_global();
+			ts = &ts_val;
+		}
```
[the remark this answers is not in the document]
Comment (rivasiv) [reply platform/linux-dpdk/odp_packet_dpdk.c:955]: I understood it's from the same burst... just same ts may cause issues in some cases. Would remove this change

**Pull request closed without merging**